src/linear_algebra/finsupp.lean

```diff
@@ -822,6 +822,63 @@ end prod
 
 end finsupp
 
+section fintype
+
+variables {α M : Type*} (R : Type*) [fintype α] [semiring R] [add_comm_monoid M] [module R M]
+variable (v : α → M)
+
+/-- `fintype.total v f` is the linear combination of vectors in `v` with weights in `f`.
+This variant of `finsupp.total` is defined on fintype indexed vectors. -/
+protected def fintype.total : (α → R) →ₗ[R] M :=
+{ to_fun := λ f, ∑ i, f i • v i,
+  map_add' := λ f g, by { simp_rw [← finset.sum_add_distrib, ← add_smul], refl },
+  map_smul' := λ r f, by { simp_rw [finset.smul_sum, smul_smul], refl } }
```

Comment:
There is a way to define `total` as a linear map unconditionally: assume that there is some type of commuting scalars using `smul_comm_class`. The result is still well-defined in the non-commutative case because any module is a `smul_comm_class` over `ℕ` and an `ℕ`-linear map is just an `add_hom`.
```suggestion
variables {α M : Type*} (R : Type*) [fintype α] [semiring R] [add_comm_monoid M] [module R M]
variables (S : Type*) [semiring S] [module S M] [smul_comm_class R S M]
variable (v : α → M)

/-- `fintype.total R S v f` is the linear combination of vectors in `v` with weights in `f`.
This variant of `finsupp.total` is defined on fintype indexed vectors.

This map is linear in `v` if `R` is commutative, and always linear in `f`.
See note [bundled maps over different rings] for why separate `R` and `S` semirings are used.
-/
protected def fintype.total : (α → M) →ₗ[S] (α → R) →ₗ[R] M :=
{ to_fun := λ v, { to_fun := λ f, ∑ i, f i • v i,
    map_add' := λ f g, by { simp_rw [← finset.sum_add_distrib, ← add_smul], refl },
    map_smul' := λ r f, by { simp_rw [finset.smul_sum, smul_smul], refl } },
  map_add' := λ u v, by { ext, simp [finset.sum_add_distrib, pi.add_apply, smul_add] },
  map_smul' := λ r v, by { ext, simp [finset.smul_sum, smul_comm _ r] } }

variables {S}
```